tests/query_params_test.go

```diff
@@ -0,0 +1,216 @@
+		t.Error("MakeRequest(): Expected handler not to be called.")
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, string(resp))
+	}
+
+}
+
+// Verifies behaviour of query parameter parser when passing malformed keys or
+// values, by breaking URL percent-encoding. Percent-encoding is used to encode
+// special characters in URL
+func TestQueryParametersBreakUrlEncoding(t *testing.T) {
+	brokenKeyReq := []byte("GET /?vegetable%=tomato HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), brokenKeyReq, func(req *http.Request) {
+		if lenQueryParams := len(req.URL.Query()); lenQueryParams != 0 {
+			t.Errorf("len(queryParams): got %d, want %d", lenQueryParams, 0)
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+
+	brokenValueReq := []byte("GET /?vegetable=tomato% HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), brokenValueReq, func(req *http.Request) {
+		if lenVeg := len(req.URL.Query()["vegetable"]); lenVeg != 0 {
+			t.Errorf("len(queryParams): got %d, want %d", lenVeg, 0)
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+
+}
+
+// Test whether both + and %20 are interpreted as space. Having a space in the
+// actual request will not be escaped and will result in a 400
+func TestQueryParametersSpaceBehaviour(t *testing.T) {
+	var (
+		value1 string
+		value2 string
+	)
+	req := []byte("GET /?vegetable=potato pear&Vegetable=tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		if req != nil {
+			t.Error("Expected the server not to receive a request containing invalid Unicode.")
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, string(resp))
+	}
+
+	req = []byte("GET /?vegetable=potato+pear&Vegetable=tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" + "\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		value1 = req.URL.Query()["vegetable"][0]
+	})
+
+	req = []byte("GET /?vegetable=potato%20pear&Vegetable=tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" + "\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		value2 = req.URL.Query()["vegetable"][0]
+	})
+
+	if value1 != value2 {
+		t.Errorf("URL.Query():want potato and potato but got %s and %s", value1, value2)
+	}
+}
```

Comment:
Consider splitting the function into 3 separate tests

Reply:
I would keep them in the same test since they are focusing on one aspect but I did include table-driven tests.